Add and rearrange documentation on getting logs (#638)
medic/cht-docs: content/en/apps/guides/debugging/obtaining-logs.md

---
title: "Obtaining Logs"
linkTitle: "Obtaining Logs"
weight:
description: >
  How to obtain server and client logs
relatedContent: >

---

There are many places where useful logs reside. This details all those places, and the easiest way to get a hold of them.

## Client logs

### On a laptop or desktop

> To check if there are relevant logs open up the developer console in your browser. The shortcut is probably COMMAND+ALT+I on MacOS, or CTRL+ALT+I on Linux and Windows. Click the console tab and copy out any errors or logging that you think is relevant.

Reviewer: ```suggestion
To check if there are relevant logs open up the developer console in your browser. The shortcut is probably COMMAND+OPTION+I on MacOS, or CTRL+SHIFT+I on Linux and Windows. Click the console tab and copy out any errors or logging that you think is relevant.
```

"ctrl + alt + I" didn't work for me, so I looked it up and [it's shift](https://developer.chrome.com/docs/devtools/shortcuts/)!

Author: Thanks! I always just hit F12...

### On a phone

There are two types of logs updated by Android devices depending on what information is needed. If you trying to get either of the logs from a device for the first time you need to set it up for USB debugging.

1. Enable USB debugging on the phone. This varies from phone to phone, but here is the [official Android documentation](https://developer.android.com/studio/debug/dev-options#enable).
2. Connect your phone by USB to the computer.

> 3. You will see a popup _"Allow USB debugging. The computer's RSA fingerprint..."_, click _"OK"_. If you don't see the popup when the cable is plugged in, you will see it after the next step.

Author: Ooof. That was an awkward copy and paste and hack job. I think the "next step" is actually step 1 in the "Android logs" section. But I don't want to encourage people to install adb just for getting the browser logs. I now think the "official instructions linked in step 1 should be enough. What do you think?

```suggestion
3. You will see a popup _"Allow USB debugging. The computer's RSA fingerprint..."_, click _"OK"_.
```

> 3. You will see a popup _"Allow USB debugging. The computer's RSA fingerprint..."_, click _"OK"_. If you don't see the popup when the cable is plugged in, you will see it after the next step.

Reviewer: agreed - this covers it enough!  And...you want me to commit? I've not had the reviewee suggest a change for the reviewer ;)

> 3. You will see a popup _"Allow USB debugging. The computer's RSA fingerprint..."_, click _"OK"_. If you don't see the popup when the cable is plugged in, you will see it after the next step.

Author: I've got it! Thanks.

   ![Allow USB debugging](/apps/guides/debugging/images/allow_usb_debugging.png)

#### Browser logs

> These are the same as the desktop browser logs. Now that you've enabled USB debugging on the phone you can access the dev console on the phone via USB from your desktop browser. Follow the [official Android documentation](https://developer.chrome.com/docs/devtools/remote-debugging/webviews/#open_a_webview_in_devtools) to access the console.

Reviewer: I totally see what you're saying, but the devils advocate in me says, "if these are the same as the desktop browser logs, why do I need them? Don't I already have them on my desktop?".  

So maybe a fix might be...

```suggestion
Now that you've enabled USB debugging on the phone you can access the dev console on the phone via USB from your desktop browser. This allows the same level of debugging and inspection you have on a desktop browser, but from your phone's browser in the CHT app. Follow the [official Android documentation](https://developer.chrome.com/docs/devtools/remote-debugging/webviews/#open_a_webview_in_devtools) to access the console.
```

Still maybe a bit awkward? Hopefully you see what I'm getting et.

#### Android logs

> The Android log is written to from the cht-android wrapper which can be useful to debug application crashes or failing integrations between apps.

Reviewer: Is it worth mentioning how these are different from the browser logs above? I ask because I want to know, so I figure others might want to know too ;)

Author: That's what this is attempting to do... I'll try again...

```suggestion
The Android log is written to from the cht-android wrapper which captures errors like application crashes or failing integrations between Android apps.
```

Better?

Reviewer: Ship it!

> 1. To **install** the `adb` command follow the instructions under the [Development Environment > Debug tool adb]({{< ref "core/guides/android/development-setup#debug-tool-adb" >}}) section.

Reviewer: ```suggestion
1. To install the `adb` command, follow the instructions under the [Development Environment > Debug tool adb]({{< ref "core/guides/android/development-setup#debug-tool-adb" >}}) section.
```

> 2. Within a command line session write the following command: `adb start-server`.

Reviewer: ```suggestion
2. Within a command line session, write the following command: `adb start-server`.
```

3. To check if your phone is properly connected, write the command `adb devices`. This will list the devices connected.

   ![ADB Devices](/apps/guides/debugging/images/adb_devices.png)

> 4. Type the command `adb logcat > phone.log` to get the android logs in the file `phone.log`. You can stop the recording at any time pressing _Ctrl+C_. Now while taking android logs is the time to try to reproduce the error. Do this for as many times as you deem necessary to capture logs for that event.

Reviewer: > Do this for as many times as you deem necessary to capture logs for that event.

Maybe call out that the log will get overwritten every time you run the command?

Author: That sentence is clunky. I think it's saying "reproduce the error" as many times as necessary, not run the command as many times as necessary. What about this...

```suggestion
4. Type the command `adb logcat > phone.log` to get the android logs in the file `phone.log`. The log will be written to the `phone.log` file as long as this command is running. You can stop the recording at any time pressing _Ctrl+C_. Now if you reproduce the error on the phone you can look for any useful information being written to the `phone.log`.
```

Reviewer: yes, more better - thanks!

This will get all the logs from the device, not just logs for the app you want to debug. You can pass arguments to `adb` to only get the logs you want.

```sh
$ adb logcat MedicMobile:V AndroidRuntime:E chromium:V '*:S' > phone.log
```

Alternatively you can use `grep` to filter the logs down to only those from the relevant app.

```sh
$ adb logcat | grep MedicMobile > phone.log
```

### On the server

> Some unexpected errors are caught and stored in `feedback` docs and stored on the phone and later synced to CouchDB on the server. To access these look for docs in the `medic-user-{username}-meta` or `medic-users-meta` databases. This is particularly useful to debug issues where you do not have physical access to the device. More information is available under [Managing Databases]({{< relref "apps/guides/database" >}}).

Reviewer: ```suggestion
Some unexpected errors are caught and stored in `feedback` docs and stored on the phone and later synced to CouchDB on the server. To access these, look for docs in the `medic-user-{username}-meta` or `medic-users-meta` databases. This is particularly useful to debug issues where you do not have physical access to the device. More information is available under [Managing Databases]({{< relref "apps/guides/database" >}}).
```

## Server logs

> To access the server logs you will need access to the server.

Reviewer: This section seems orphaned now that we have removed the link to  [cht-logs documentation](https://github.com/medic/cht-conf#user-content-fetching-logs) and now that we have the above "On the server" section.  I propose removing these two lines.

Reviewer: This looks still orphaned in the latest commit on this branch

> To access the server logs you will need access to the server.

Author: Wow, you're right, this is terribly redundant. It'll also change a little with CHT v4. The reason I dropped the documentation on the cht-logs command is it only worked on v2.

I'm tempted to drop this altogether until we have something useful to put here. What do you think? If I do that do I have to rename this whole file to be client centric (ie: drop the word server from everywhere)?

Reviewer: Yup - I say drop it all together

Author: Done
